### PR TITLE
Implement non-greedy tokenizer that tries to maximize token lengths

### PR DESCRIPTION
**Pull request by thement**

### CODE REVIEW

File: utils.cpp [hunk 1]
```diff
@@ -275,41 +275,57 @@ std::vector<gpt_vocab::id> gpt_tokenize(const gpt_vocab & vocab, const std::stri
     return tokens;
 }
 
+// TODO: Calculate this constant from the vocabulary
+#define MAX_TOKEN_LEN 10
```
Comment (j-f1) [inline utils.cpp:279]: Max is actually 16!

```
▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
----------------
▁Einzelnachweise
****************
================
################
addEventListener
////////////////
▁implementations
▁straightforward
▁characteristics
▁representations
................
________________
▁Representatives
▁transformations
```

There are also a bunch more:


<details>
<summary>15</summary>

```
▁implementation
springframework
includegraphics
▁representation
▁authentication
▁classification
▁transformation
OnClickListener
printStackTrace
▁characteristic
▁administration
▁initialization
▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
▁estadounidense
▁administrative
▁Unterscheidung
▁interpretation
▁internationale
▁representative
▁multiplication
▁simultaneously
▁principalmente
▁configurations
▁responsibility
▁Classification
▁Administration
▁Fußballspieler
▁Schriftsteller
▁veröffentlicht
▁setContentView
▁Authentication
▁Mediabestanden
PropertyChanged
▁recommendation
backgroundColor
▁identification
```

</details>


<details>
<summary>14</summary>

```
▁▁▁▁▁▁▁▁▁▁▁▁▁▁
▁International
▁documentation
▁configuration
▁international
▁automatically
▁corresponding
ViewController
getElementById
▁understanding
▁functionality
▁Unfortunately
▁communication
AndroidRuntime
▁circumstances
▁approximately
▁Championships
▁unfortunately
ActivityThread
eltemperaturen
▁Massachusetts
Authentication
▁approximation
▁significantly
▁Internacional
▁Bibliographie
amerikanischer
▁Entertainment
disambiguation
▁distributions
▁consideration
▁universitaire
▁internacional
▁Alternatively
▁autorytatywna
▁distinguished
▁contributions
▁Configuration
adratkilometer
implementation
▁relationships
▁specification
preventDefault
▁investigation
▁partiellement
Longrightarrow
authentication
▁административ
▁contradiction
longrightarrow
▁développement
▁possibilities
▁compatibility
▁Transfermarkt
▁collaboration
▁establishment
▁verschiedenen
▁independently
▁organizations
▁StringBuilder
▁notifications
▁decomposition
▁modifications
LayoutInflater
RelativeLayout
CollectionView
▁concentration
▁computational
▁administrator
▁participation
▁extraordinary
▁französischen
▁Genomsnittlig
▁authorization
▁environmental
▁stackoverflow
▁neighbourhood
▁interpolation
```

</details>

I am a little bit worried we may run into “glitch tokens” like GPT-3 (but that’s something that can be explored in the future!)

**Review by ggerganov — APPROVED** [review]
Merge it if results look ok.
I won't be able to have detailed look in the next few days

File: utils.cpp (hunk 1 above)
Comment (thement) [reply utils.cpp]: Thanks! I didn't realize there were tokens like that . I changed it to 18 with intention to calculate this from the vocabulary in the future.

File: main.cpp [hunk 2]
```diff
@@ -846,6 +846,7 @@ int main(int argc, char ** argv) {
     std::vector<float> logits;
 
     // tokenize the prompt
+    params.prompt.insert(0, 1, ' ');
```
[the remark this answers is not in the document]
Comment (thement) [reply main.cpp:849]: It should be fused to the first token! This is how original python llama code parses it.

**Pull request merged**